src/persist/src/indexed/encoding.rs

```diff
@@ -0,0 +1,564 @@
+#[derive(Clone, Debug, Abomonation)]
+pub struct BlobTraceMeta {
+    /// The batches that make up the BlobTrace, represented by their description
+    /// and the key to retrieve the batch's data from the blob store. Note that
+    /// Descriptions are half-open intervals `[lower, upper)`.
+    pub batches: Vec<(Description<u64>, String)>,
+}
+
+/// The structure serialized and stored as a value in [crate::storage::Blob]
+/// storage for data keys corresponding to future data.
+///
+/// Invariants:
+/// - The seqno of each update is >= to desc.lower() and < desc.upper().
+/// - The values in updates are sorted by (time, key, value).
+/// - The values in updates are "consolidated", i.e. (time, key, value) is
+///   unique.
+/// - All entries have a non-zero diff.
+#[derive(Clone, Debug, Abomonation)]
+pub struct BlobFutureBatch {
+    /// Id of the stream this batch belongs to.
+    pub id: u64,
+    /// Which updates are included in this batch.
+    pub desc: Description<SeqNo>,
+    /// The updates themselves.
+    pub updates: Vec<(SeqNo, (String, String), u64, isize)>,
+}
+
+/// The structure serialized and stored as a value in [crate::storage::Blob]
+/// storage for data keys corresponding to trace data.
+///
+/// Invariants:
+/// - The timestamp of each update is >= to desc.lower() and < desc.upper().
+/// - The values in updates are sorted by (key, value, time).
+/// - The values in updates are "consolidated", i.e. (key, value, time) is
+///   unique.
+/// - All entries have a non-zero diff.
+///
+/// TODO: This probably wants to be a different level of abstraction, so we can
+/// put multiple small batches in a single blob but also break a very large
+/// batch over multiple blobs. We also may want to break the latter into chunks
+/// for checksum and encryption?
+#[derive(Clone, Debug, Abomonation)]
+pub struct BlobTraceBatch {
+    /// Id of the trace this batch belongs to.
+    pub id: u64,
+    /// Which updates are included in this batch.
+    pub desc: Description<u64>,
+    /// The updates themselves.
+    pub updates: Vec<((String, String), u64, isize)>,
+}
+
+impl Default for BlobMeta {
+    fn default() -> Self {
+        BlobMeta {
+            last_file_id: Self::now_millis(),
+            futures: Vec::new(),
+            traces: Vec::new(),
+        }
+    }
+}
+
+impl BlobMeta {
+    /// Asserts Self's documented invariants, returning an error if any are
+    /// violated.
+    pub fn validate(&self) -> Result<(), Error> {
+        for (_, f) in self.futures.iter() {
+            f.validate()?;
+        }
+        for (_, t) in self.traces.iter() {
+            t.validate()?;
+        }
+        Ok(())
+    }
+
+    /// Returns the current SystemTime in millis since epoch.
+    pub fn now_millis() -> u128 {
+        SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .unwrap_or_default()
+            .as_millis()
+    }
+}
+
+impl BlobFutureMeta {
```

Comment:
I'm not sure if theres something we want to assert with `ts_lower` and the rest of the data in this struct or not. I don't actually think so (ts_lower can be arbitrarily far back or forward relative to the data contained in `self.batches` afaict but just wanted to call it out)

Reply:
Maybe theres an invariant that `ts_lower` is equivalent to the greatest `upper` for the corresponding trace?